include/logging/log_msg2.h

```diff
@@ -394,6 +373,36 @@ do { \
 	} \
 	(void)_mode; \
 } while (0)
+
+/* Macro for getting name of a local variable with the exception of the first argument
+ * which is a formatted string in log message.
+ */
+#define Z_LOG_LOCAL_ARG_NAME(idx, arg) COND_CODE_0(idx, (arg), (_v##idx))
+
+/* Create local variable from input variable (expect first (fmt) argument). */
+#ifdef __cplusplus
+#define Z_LOG_LOCAL_ARG_CREATE(idx, arg) \
+	COND_CODE_0(idx, (), (auto Z_LOG_LOCAL_ARG_NAME(idx, arg) = (arg) + 0))
+#else
+#define Z_LOG_LOCAL_ARG_CREATE(idx, arg) \
+	COND_CODE_0(idx, (), (__auto_type Z_LOG_LOCAL_ARG_NAME(idx, arg) = (arg) + 0))
```

Comment:
Can we have a define that evaluates to either auto or __auto_type so this is simplified ?

Reply:
Done, Setting `Z_AUTO_TYPE` define to `auto` in cpp and `__auto_type` if c (and supported).